pages/implementers/interfaces.md

```diff
@@ -0,0 +1,203 @@
+
+## Instance Validation
+
+Instance validation is one of the key capabilities of JSON Schema.
+It is the process in which a given piece of data is deemed valid or invalid under a specific schema.
+Implementations may offer one or more of the specific interfaces below in order to perform this validation.
+
+### Exception-Driven Validation
+
+**Type**: `EvaluationOptions → None <!> ValidationError` or `EvaluationOptions → Result <!> ValidationError`
+
+A validation API which causes a language-specific failure or exception when validation itself fails.
+If it succeeds, this API may return a result with further detail, or may simply continue execution silently.
+
+### Boolean Validation
+
+**Type**: `EvaluationOptions` → Boolean
+
+An API which produces a simple boolean result indicating an instance's validity under a schema.
+
+### Two-Argument Validation
+
+**Type**: `Schema` × `Instance` → `Result`
+
+An API which takes a schema and instance simultaneously and produces a result indicating whether the instance is valid under the given schema.
+
+In some sense, two argument validation is the simplest possible API for JSON Schema validation; it asks simply whether a given instance is valid under a given schema with no presumptions of repeated use (of the schema) or additional calculation.
+
+### Repeated Validation / Schema Compilation
+
+**Type**: `Schema` → `Instance` → `Result`
+
+An API which attempts to prepare a schema for repeated use.
+It may (and likely will) perform some form of preoptimization, performing some set of work ahead of time such that it will not be necessary to repeat when validating many instances.
+
+## Annotation Collection
+
+**Type**: `EvaluationOptions` → `ResultWithAnnotations`
+
+An API which collects annotations produced when processing a given schema and instance.
+
+## Schema Validation
+
+**Type**: `Schema` → `Result`
+
+An API which validates a schema itself under the dialect it is written for.
+This API likely makes use of a corresponding metaschema (or metaschemas) for the dialect, but generally must do additional work to ensure the schema is not invalid even under conditions not checked for within the metaschema.
+
+## Explicit Version Selection
+
+**Type**: `Dialect` → `EvaluationOptions` → `Result`
+
+An API which controls which dialect the implementation will assume when given a schema which does not otherwise indicate its dialect (i.e. which does not declare a `$schema` property).
+
+## Version Detection
+
+**Type**: `Schema` → `Dialect`
+
+An API which identifies which dialect a given schema is written for, returning either the dialect itself or otherwise a dialect-specific validation function.
+
+## Type Customization
+
+An API which allows (re-)configuring which language-level types correspond to which JSON Schema types, and additionally potentially allowing for the definition of new types.
```

Comment:
By this do you mean, like having a consturctor for the library so it can be used in multiple configurations simultaneously? Or, things can be re-configured? Or both?

Reply:
Whether it's per evaluation or once globally I suppose I didn't specify here (nor for other sections like the schema registry -- that too may either be something global or per-evaluation seemingly).

But the point is/was, if your programming language has multiple container types (e.g. a dynamic array, a linked list, and an immutable list, say) how do you configure which of those types map to the JSON `array` type (and therefore to JSON Schema's `items` keyword, say.)

Reply:
> Whether it's per evaluation or once globally I suppose I didn't specify here (nor for other sections like the schema registry -- that too may either be something global or per-evaluation seemingly).

Ah, OK. I don't know that this element specifically matters in relation to JSON Schema. Although, I have seen problems where a library is unable to run multiple differnt dialects within the same file/scope. I realise this wasn't what you were getting at. I think there's globaly, per-evaluation, and per instance of a library (inbetween the two), which could apply to all configuration options and/or a schema registry. I could see the value in having both a global and instance based schema registry.

> But the point is/was, if your programming language has multiple container types (e.g. a dynamic array, a linked list, and an immutable list, say) how do you configure which of those types map to the JSON `array` type (and therefore to JSON Schema's `items` keyword, say.)

Oh, I'd never even considered that. My experience with typed langauges is limited, but is that sort of dynamic deserialization something that's common? (If I'm understanding this correctly, that's what you're suggesting?). I Imagin such would be quite tricky?

Reply:
(Agreed on whether you have global state or not not being very interesting, that's why I didn't really mention it.)

---
> >  My experience with typed langauges is limited, but is that sort of dynamic deserialization something that's common?

My point of reference for this is my own implementation, where it looks like:

```
from jsonschema import validators
class MyInteger:
    ...
type_checker = Draft202012Validator.TYPE_CHECKER.redefine(
  "number",
  lambda _, value: Draft202012Validator.TYPE_CHECKER.is_type(value, "number") or isinstance(value, MyInteger),
)

CustomValidator = validators.extend(
    Draft202012Validator,
    type_checker=type_checker,
)
validator = CustomValidator(schema={"type" : "number"})
# instances of `MyInteger` are going to be treated as numbers
```

I don't know what kinds of languages are the ones where this would or wouldn't happen actually -or how common it is in implementations -- I'd personally have thought it'd be found in some as certainly in Python it's very long practice that JSON users want to extend it with new types (e.g. "how do I write a JSON array in JSON but then I want that to deserialize into a set in my programming language not an array"). Can look around though to see what other implementations have done in JSON Schema land.

Reply:
> it sounds like you're saying there's an approach for saying "hey, you can cast this thing to a number" and processing native types as JSON for JSON Schema driven validation?

I think so :D -- I mean I'm somewhat confident I'm saying something concrete and that your rephrasing is the same thing, so I think so -- I think the vast *majority* of implementations do what you're calling "processing native types as JSON" if I parse those words correctly. A minority do "I'll parse a string into some types on your behalf" (which is represented on this page with the interface for "String based validation").

Reply:
I think I'm on the same page now. I guess I just forget about these sorts of internals when I mostly work in javascript/node.js land.